packages/lodestar-fork-choice/src/forkChoice/errors.ts

```diff
@@ -174,7 +174,7 @@ export class ForkChoiceError extends Error {
   type: ForkChoiceErrorType;
 
   constructor(type: ForkChoiceErrorType) {
-    super();
+    super(type.code + " " + JSON.stringify(type));
```

Comment:
I tend to think this JSONification should either happen downstream, eg: in logging, or pre-logging, or we should have a custom `toString` method here. Perhaps this can be tackled in a separate PR.

Reply:
I feel like this is specific to error and to respect error message property, it should be handled here like mentioned in https://github.com/ChainSafe/lodestar/pull/1579#discussion_r493242136

Reply:
The problem with mashing the stringified contents here is that you end up with a bunch of duplicate information (and the duplicated info is nasty, stringified and inflexible to different formatting.) This goes against the goal of https://github.com/ChainSafe/lodestar/issues/1249:
```
{
  message: "{\"code\":\"ERR_FUTURE_SLOT\",\"blockSlot\":5,\"currentSlot\":4}",
  stack: "...",
  type: {
    code: "ERR_FUTURE_SLOT",
    blockSlot: 5,
    currentSlot: 4,
  }
}
```

My current thought is that we should provide a custom method on our errors, eg: `toObject`, or `toJson`, and our logger should attempt to convert errors to objects, and do the stringification.  Would have something like the following logic:
```typescript
function defaultErrorObj(e: Error): object {
  const obj = {};
  e.message && obj.message = e.message;
  e.code && obj.code = e.code;
  e.stack && obj.stack = e.stack;
  return obj;
}

function convertError(e: Error): object {
  if (e.toObject) {
    return e.toObject(); // custom errors can supply metadata here
  } else {
    return defaultErrorObj(e); // regular errors handled here
  }
}

// alternatively, substitute `object` for `Json`
```
This allows each specific error to provide its own relevant object-ification, but lets the actual string-ification be deferred to the level of logging, which knows better how strings should be formatted.

Reply:
I honestly don't think we should have metadata on Error and we should respect existing Error interface of message/stacktrace.
We can just as easily user logger serialization to serialize metadata into error message field which should work nicely with our logger. I don't think we are using error metadata anywhere in runtime?!

Reply:
> I honestly don't think we should have metadata on Error and we should respect existing Error interface of message/stacktrace.

The two aren't mutually exclusive. We can have metadata on an `Error` subclass, and it will respect the superclass interface (message/stacktrace). If we really want to respect the interface, semantically, we should just always pass in a string description of the error (the `message`).

Error metadata gives error handlers _actual data_ to handle error scenarios with more sophistication.

Eg: Do you want an error handler to handle `ERR_FUTURE_SLOT`, but only when the `blockSlot` is within 5 of the `currentSlot`?

If you rely on `err.message` only, thats just too bad, your errors are simply not expressive enough. Or perhaps the handler must now perform some error prone string parsing or regex. If you have error metadata, this becomes possible, and trivial.

> We can just as easily user logger serialization to serialize metadata into error message field which should work nicely with our logger.

Its actually not just as easy. The reason is that string representation requires _serialization_ of structured data. If you preserialize Error metadata, you're literally recreating the situation we had before the logger context object was available!

The structured logger initiative is about _retaining_ structured data, and having the logger perform the serialization that is necessary. By jamming structured data into the error.message you remove the ability of the logger to do its own serialization. 

> I don't think we are using error metadata anywhere in runtime?!

We likely will very soon :) we've only just integrated these (and `BlockError`).

Reply:
Error subclass (with message, stack) + metadata provides the best of both worlds. Having some somewhat expressive message is useful too in case the error leaks to unexpected places, so it's more debuggable.

Reply:
@wemeetagain @mpetrunic would it be best to open a discussion issue and agree on the Lodestar Error strategy?

Reply:
> @wemeetagain @mpetrunic would it be best to open a discussion issue and agree on the Lodestar Error strategy?

yup